### ISSUE TITLE
External image proxy alternatives?

### ISSUE THREAD

**Issue by PeterDaveHello** [mattermost/mattermost-server#17357]
Sorry that this issue doesn't fit the original issue template.

I know that Mattermost suggests [atmos/camo](https://github.com/atmos/camo/) as external image proxy here: https://docs.mattermost.com/administration/image-proxy.html#atmos-camo-image-proxy, but looks like it's not being maintained for a few years, its dated dependency could have so many critical security issues.

Not sure if Mattermost would like to suggest any other image proxy that can be easily integrated with, or even help maintain [atmos/camo](https://github.com/atmos/camo/)(or fork it, and maintain Mattermost's version)?

Thanks for your time and effort for maintaining Mattermost and reading this issue :)

**Comment by amyblais** [comment]
@hmhealey ^

**Comment by PeterDaveHello** [comment]
It's now archived, see:
https://github.com/atmos/camo/issues/145#issuecomment-814948294

and the mentioned alternative: https://github.com/cactus/go-camo

**Comment by hmhealey** [comment]
Thanks for the heads up! I hadn't heard about that before now since I haven't been working with our image proxy for a while either. I don't know if we have anyone actually maintaining that right now, so I'll bring it up with the team.

As for alternatives, https://github.com/cactus/go-camo sounds like it might even work as a drop-in replacement for atmos/camo as well. I haven't tried it out yet, but it sounds like they tried to match camo's implementation of the parts we used. That one sounds promising.

The person who implemented it originally did want to support https://github.com/willnorris/imageproxy, so there might be code around for that in an old branch. That's also used by the internal image proxy that I added later, so I assume it's safe at least.

**Comment by thomas-mc-work** [comment]
Is it required to have a certain proxy implementation here? Why isn't it possible to have a generic interface to use any solution?

**Comment by thomas-mc-work** [comment]
What about [the default HTTP proxy mechanism used by all browsers](https://en.wikipedia.org/wiki/Proxy_server#Web_proxy_servers)? That would make choosing an implementation as easy as possible.

**Comment by zetaab** [comment]
![image](https://user-images.githubusercontent.com/22482503/142639909-1e32f715-e5ad-4fd8-9362-88bcba148679.png)

should we start doing something?

**Comment by hmhealey** [comment]
I've raised this to our security team to see if we can get more attention on this. This does seem like something we should be looking into, but unfortunately, even though I've worked on that code before, I'm not able to look at it right now due to other responsibilities.

**Comment by jupenur** [comment]
@zetaab the scan results in the screenshot seem to be from `trivy image` -- which means the vulnerabilities are not specific to `atmos/camo` but to the Docker image, which is based on `node:8.4`.

**You should not be using the Docker image without updating it first!**

In fact, you should probably not be using the Docker image at all. The `README.md` file clearly says: **You should run this on heroku.**

The proxy itself is very small, under 300 LoC with 0 external dependencies. As far as I can tell, there are no publicly known vulnerabilities in the proxy code. If you deploy it properly, using a recent node version, you'll have no security issues.

That said, since there is an actively maintained alternative at [cactus/go-camo](https://github.com/cactus/go-camo#differences-from-camo), that's what everybody should *really* be using. I haven't tested it, but since the only documented backwards-incompatibility is the lack of support for for the "Query String Format", it should work as a drop-in replacement without any changes in Mattermost. [Mattermost uses the path format](https://github.com/mattermost/mattermost-server/blob/dfa476a6882373ba73b429fd9d9ea796de28bd41/services/imageproxy/atmos_camo.go#L91), which `go-camo` supports.

---

TL;DR: using `atmos/camo` is not a security issue as long as you don't use an outdated Docker image, but really you should migrate over to `cactus/go-camo` instead if you can.

**Comment by dropwhile** [comment]
If you are worried about robustness and/or usage, go-camo is known to be used by:
* [gitlab](https://docs.gitlab.com/ee/security/asset_proxy.html)
* [zulip](https://zulip.readthedocs.io/en/latest/production/security-model.html#user-uploaded-content-and-user-generated-requests)

and several others.

If you have any questions, feel free to reach out.
